internal/validators/namespace_test.go

```diff
@@ -222,38 +228,49 @@ func setSubAnnotation(ns *corev1.Namespace, pnm string) {
 	ns.SetAnnotations(a)
 }
 
-func TestIllegalExcludedNamespace(t *testing.T) {
+func TestIllegalIncludedNamespaceNamespace(t *testing.T) {
 	f := foresttest.Create("-a-c") // a <- b; c <- d
 	vns := &Namespace{Forest: f}
-
-	// Non-excluded namespaces with excluded-namespace label are illegal.
-	illegalInst := &corev1.Namespace{}
-	illegalInst.Name = "illegal"
-	illegalInst.SetLabels(map[string]string{api.LabelExcludedNamespace: "true"})
-	legalInst := &corev1.Namespace{}
-	legalInst.Name = "legal"
-	legalInst.SetLabels(map[string]string{api.LabelExcludedNamespace: "true"})
-	config.ExcludedNamespaces["legal"] = true
+	config.ExcludedNamespaces["excluded"] = true
 
 	tests := []struct {
-		name   string
-		nsInst *corev1.Namespace
-		op     k8sadm.Operation
-		fail   bool
+		name       string
+		excludedNS bool
+		lval       string
+		op         k8sadm.Operation
+		fail       bool
 	}{
-		{name: "allow updating legal", nsInst: legalInst, op: k8sadm.Update},
-		{name: "allow creating legal", nsInst: legalInst, op: k8sadm.Create},
-		{name: "allow deleting legal", nsInst: legalInst, op: k8sadm.Delete},
-		{name: "deny updating illegal", nsInst: illegalInst, op: k8sadm.Update, fail: true},
-		{name: "deny creating illegal", nsInst: illegalInst, op: k8sadm.Create, fail: true},
-		{name: "allow deleting illegal", nsInst: illegalInst, op: k8sadm.Delete},
+		{name: "allow creating non-excluded namespace without label", op: k8sadm.Create},
+		{name: "allow creating non-excluded namespace with label (e.g. from a yaml file)", lval: "true", op: k8sadm.Create},
+		{name: "deny creating non-excluded namespace with label with wrong value(e.g. from a yaml file)", lval: "else", op: k8sadm.Create, fail: true},
+		{name: "allow creating excluded namespace without label", excludedNS: true, op: k8sadm.Create},
+		{name: "deny creating excluded namespace with label", excludedNS: true, lval: "true", op: k8sadm.Create, fail: true},
+
+		{name: "allow updating non-excluded namespace with correct label (which is actually set by HC reconciler)", lval: "true", op: k8sadm.Update},
+		{name: "deny updating non-excluded namespace without label", op: k8sadm.Update, fail: true},
```

Comment:
Does this mean that if I say "kubectl apply -f ns.yaml" twice on the same file, it could pass the first time (if the ns doesn't exist yet) and fail the second time? That doesn't really seem right. Lots of people and systems (include Config Sync) automatically say "apply -f" and expect it to work, and if the config didn't know about our labels, this will break them.

Maybe we should just allow it and trust the reconciler to fix things up? It does open up a very brief security window but perhaps we should plug that with a mutating admission webhook, not a validating one. Maybe file an issue to turn this into a mutating webhook?

Reply:
Agreed. Updated to allow and filed #37 .